src/pymor/reductors/neural_network.py

```diff
@@ -16,10 +18,92 @@
     from pymor.algorithms.pod import pod
     from pymor.core.base import BasicObject
     from pymor.core.exceptions import NeuralNetworkTrainingFailed
-    from pymor.models.neural_network import (FullyConnectedNN,
-                                             NeuralNetworkModel,
-                                             NeuralNetworkInstationaryModel,
-                                             NeuralNetworkOutputModel)
+    from pymor.algorithms.neural_network import (
+            FullyConnectedNN, EarlyStoppingScheduler, CustomDataset, certified_ann_training)
+    from pymor.models.neural_network import (
+            StationaryNeuralNetworkModel, NeuralNetworkModel, NeuralNetworkInstationaryModel, NeuralNetworkOutputModel)
+    from pymor.parameters.base import Mu
+    from pymor.reductors.basic import StationaryRBReductor
+    from pymor.vectorarrays.interface import VectorArray
+
+
+    class StationaryNeuralNetworkRBReductor(StationaryRBReductor):
+        '''
+        assumes RB to be orthogonal w.r.t. product
+        '''
+
+        def __init__(self, fom, training_data, validation_data,
+                RB=None, product=None, check_orthonormality=None, check_tol=None,
+                hidden_layers='[(N+P)*3, (N+P)*3]', ann_mse='like_basis',
+                max_restarts=10, torch_seed=None, ann_train_params=None):
```

Comment:
The nice thing about the `NeuralNetworkReductor` was that one could call it with different parameters. For instance, if the loss obtained with a trained network was not sufficient, one could have changed the number of hidden layers in an automatic routine to (hopefully) improve the training results. When having to pass all those information to `__init__`, this is not possible anymore. However, I understand that this is required here, if we do not want to override the `reduce` method of `ProjectionBasedReductor`.

Reply:
I am for instance thinking about a greedy algorithm which calls `reduce()` on the same reductor in each iteration after extending the basis. There, we do nat want the greedy to need to know about the kind of reductor. However, we now have the surrogates as input for the greedy which are there for exactly that reason. So, its not that clear. We should probably we involve the other pyMOR devs in the discussion at some point... :)

Reply:
And the neural network architecture would be adjusted whenever calling `reduce()` since the required output dimension of the neural network might have changed, right?

Reply:
Exactly, since the RB and its size changed. But this could also mean that a different number of restarts is appropriate for different reduce calls. So its really not fully clear to me where to put things...

Reply:
Yes, that makes everything a bit difficult. For instance, the architecture of the neural network should be adjusted as well when increasing the number of outputs in that case (at least in general). Hence, the original neural network architecture and training procedure might be inappropriate when changing the training data. Some time ago, I discussed with @sdrave whether it is possible to implement some sort of general "fitting" routine that adjusts neural network architecture and training parameters automatically to obtain a sufficient model. But I think this is quite difficult... And such a routine might increase the training time dramatically.